File: projects/02-rate-limiter/src/token_bucket.rs
//! V1 — Token bucket, built from scratch (in-process, single node).
//!
//! This is the algorithm in its purest form: a pure struct with no I/O, easy to
//! unit-test exhaustively before you ever involve Redis. Get this right and V3
//! becomes "the same math, but atomic and in Lua".
//!
//! The defining trick: **refill lazily, on read.** There is no background timer
//! ticking tokens into a million buckets. Instead, every `try_acquire` computes
//! how many tokens *would have* accrued since `last_refill` from the elapsed
//! time, caps at capacity, and only then tries to spend.

use std::time::Instant;

use crate::limiter::{Decision, LimitConfig};

/// A single key's token bucket. Cheap; one per key.
pub struct TokenBucket {
    /// Bucket capacity — the most tokens you can ever hold (the burst).
    capacity: f64,
    /// Tokens added per second (the sustained rate).
    refill_per_sec: f64,
    /// Current token count. Fractional on purpose — don't round away budget.
    tokens: f64,
    /// When `tokens` was last brought up to date.
    last_refill: Instant,
}

impl TokenBucket {
    /// A fresh bucket starts **full** (allowed to burst immediately).
    pub fn new(cfg: LimitConfig) -> Self {
        Self {
            capacity: cfg.burst as f64,
            refill_per_sec: cfg.rate_per_sec,
            tokens: cfg.burst as f64,
            last_refill: Instant::now(),
        }
    }

    /// Account for a request costing `cost` tokens, as of `now`.
    ///
    /// `now` is injected (rather than read internally) so tests can drive time
    /// deterministically.
    pub fn try_acquire(&mut self, cost: u64, now: Instant) -> Decision {
        // TODO(V1): implement lazy refill + acquire.
        //   1. Refill: add `elapsed.as_secs_f64() * refill_per_sec` tokens,
        //      clamped to `capacity`, then advance `last_refill` to `now`.
        //   2. If `tokens >= cost`: deduct and return `Decision::allow(...)`.
        //   3. Else: compute how long until `cost` tokens exist and return
        //      `Decision::deny(retry_after, capacity)`.
        // Mind the precision: keep `tokens` fractional; never let rounding
        // manufacture or destroy budget across many calls.
        let _ = (
            cost,
            now,
            self.capacity,
            self.refill_per_sec,
            self.tokens,
            self.last_refill,
        );
        todo!("V1: lazy token-bucket refill + acquire")
    }
}

#[cfg(test)]
mod tests {
    // TODO(V1): prove the algorithm. Suggested cases:
    //   - a full bucket allows exactly `burst` immediate requests, then denies;
    //   - after waiting `1/rate` seconds, exactly one more is allowed;
    //   - tokens never exceed capacity no matter how long you wait;
    //   - `retry_after` on a deny is truthful (waiting it out then allows).
    // Use an injected `Instant` base + `base + Duration::from_*` to control time.
}

File: projects/02-rate-limiter/src/rate_limiter/token_bucket.py
"""V1 — Token bucket, built from scratch (in-process, single node).

The algorithm in its purest form: a small class with no I/O, which you can
unit-test exhaustively before Redis is anywhere near it. Get this right and V3
becomes "the same arithmetic, but atomic and written in Lua".

The defining trick is **lazy refill — compute on read.** There is no background
task ticking tokens into a million buckets; a task per key would cost more than
the service it protects. Instead every `try_acquire` asks "how many tokens
*would* have accrued since I last looked?", caps that at capacity, and only then
tries to spend.

**The Python trap.** `tokens` is a `float`, and floats are why "we allow 10/sec"
quietly becomes 9.9997/sec after a few million requests. Two specific hazards:

  * Adding `elapsed * rate` a few million times accumulates representation
    error. Whether that matters depends on whether you also *clamp* at capacity
    — think about which operations lose budget and which manufacture it.
  * `int(x)` truncates toward zero and `round()` uses banker's rounding. Neither
    is obviously right here. Decide what you are rounding, and where — the SPEC
    property-tests that the long-run rate does not drift.

Scaffold state: the bucket constructs; the first `try_acquire` raises. That is
your worklist.
"""

from __future__ import annotations

from .limiter import Decision, LimitConfig

__all__ = ["TokenBucket"]


class TokenBucket:
    """One key's token bucket. Cheap — expect one per active key.

    A fresh bucket starts **full**, so a caller that has been quiet may burst up
    to `capacity` immediately. That is the point of having a burst knob at all.
    """

    __slots__ = ("_capacity", "_last_refill", "_refill_per_sec", "_tokens")

    def __init__(self, config: LimitConfig, now: float) -> None:
        self._capacity = float(config.burst)
        self._refill_per_sec = config.rate_per_sec
        self._tokens = float(config.burst)
        self._last_refill = now

    @property
    def tokens(self) -> float:
        """Tokens as of the last refill — for tests and `/metrics`, not decisions.

        Deliberately *not* refreshed on read: a property that silently advances
        time would make the drift property test meaningless.
        """
        return self._tokens

    def try_acquire(self, cost: int, now: float) -> Decision:
        """Account for a request costing `cost` tokens, as of monotonic `now`.

        `now` is a parameter rather than a `time.monotonic()` call inside, so a
        test can drive a thousand refills across a simulated hour instantly.
        """
        # TODO(V1): lazy refill, then acquire.
        #   1. Refill: add `(now - self._last_refill) * self._refill_per_sec`
        #      tokens, clamp to `self._capacity` (`min` is the whole clamp), and
        #      advance `self._last_refill` to `now`.
        #   2. If there are at least `cost` tokens: deduct and
        #      `return Decision.allow(...)`.
        #   3. Otherwise work out how long until `cost` tokens exist — a
        #      division, and think about what it means when the rate is 0 — and
        #      `return Decision.deny(retry_after, ...)`.
        #
        # Keep `self._tokens` a float the whole way through. `remaining` on the
        # Decision is an int because it goes on the wire; converting there (not
        # in the state) is what stops rounding from eating the budget.
        raise NotImplementedError("V1: lazy token-bucket refill + acquire")

    def peek(self, now: float) -> Decision:
        """Report state as of `now` **without** consuming budget.

        Backs the `Peek` RPC. Note the subtlety: refilling is not "consuming",
        so a peek may legitimately advance `_last_refill` — but it must never
        change the answer a concurrent `try_acquire` would have given.
        """
        # TODO(V1): refill (or compute the refilled value without storing it —
        # decide which, and say why) and report `remaining` and a `retry_after`
        # for the caller's "am I about to be limited?" probe.
        raise NotImplementedError("V1: non-consuming peek at bucket state")
